Microsoft.O365.Security.Native.ETW/Callbacks.hpp

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include <krabs.hpp>
+
+#include "EventRecord.hpp"
+#include "EventRecordError.hpp"
+
+using namespace System;
+using namespace System::Runtime::InteropServices;
+
+namespace Microsoft { namespace O365 { namespace Security { namespace ETW {
+
+    /// <summary>
+    /// Delegate called when a new ETW <see cref="O365::Security::ETW::EventRecord"/> is received.
+    /// </summary>
+    public delegate void IEventRecordDelegate(IEventRecord^ record);
+
+    /// <summary>
+    /// Delegate called when a new ETW <see cref="O365::Security::ETW::EventRecordMetadata"/> is received.
+    /// </summary>
+    public delegate void IEventRecordMetadataDelegate(IEventRecordMetadata^ record);
+
+    /// <summary>
+    /// Delegate called on errors when processing an <see cref="O365::Security::ETW::EventRecord"/>.
+    /// </summary>
+    public delegate void EventRecordErrorDelegate(IEventRecordError^ error);
+
+    delegate void EventNativeDelegate(const EVENT_RECORD&, const krabs::trace_context&);
+
+    delegate void EventErrorNativeDelegate(const EVENT_RECORD&, const std::string&);
+
+    ref class CallbackBridge {
+    private:
+
+        EventNativeDelegate^ eventDelegateKeepAlive_;
+        EventErrorNativeDelegate^ errorDelegateKeepAlive_;
+        EventRecordMetadata^ metadata_;
```

Comment:
NB: these instances are reused which reduces per-event allocations.

Reply:
It seems that these objects have been reused.

``` cpp
if (!eventDelegateKeepAlive_)
    eventDelegateKeepAlive_ = gcnew EventNativeDelegate(this, &CallbackBridge::EventNotification);
```